plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/DefaultHttpResponseSender.java

```diff
@@ -50,62 +61,180 @@
-            if (hashCode != null && !httpResponse.containsHeader(HttpHeaders.ETAG)) {
-              httpResponse.setHeader(
-                  HttpHeaders.ETAG,
-                  hashCode.toString()
-              );
-            }
+        if (request != null && !HttpMethods.HEAD.equals(request.getAction())) {
+          try (InputStream input = payloadResponse.getPayload().openInputStream()) {
+            sendContent(input, httpResponse);
           }
         }
+      }
+    }
+    else {
+      httpResponse.sendError(status.getCode(), status.getMessage());
+    }
+  }
 
-        if (request != null && !HttpMethods.HEAD.equals(request.getAction())) {
-          try (InputStream input = payload.openInputStream(); OutputStream output = httpResponse.getOutputStream()) {
-            ByteStreams.copy(input, output);
+  /**
+   * Set HTTP response headers based on {@link Response}.
+   */
+  protected void handleResponseHeaders(final Response response,
+                                       final HttpServletResponse httpResponse)
+  {
+    // add response headers
+    for (Map.Entry<String, String> header : response.getHeaders()) {
+      httpResponse.addHeader(header.getKey(), header.getValue());
+    }
+  }
+
+  /**
+   * Set HTTP response headers based on {@link PayloadResponse}.
+   */
+  protected void handlePayloadHeaders(final Payload payload,
+                                      final HttpServletResponse httpResponse)
+  {
+    if (payload.getContentType() != null) {
+      httpResponse.setContentType(payload.getContentType());
+    }
+    if (payload.getSize() != Payload.UNKNOWN_SIZE) {
+      httpResponse.setContentLengthLong(payload.getSize());
+    }
+
+    if (payload instanceof DecoratedPayload) {
+      final DecoratedPayload decoratedPayload = (DecoratedPayload) payload;
+      handleDecoratedPayloadHeaders(decoratedPayload, httpResponse);
+    }
+  }
+
+  /**
+   * Set HTTP response headers based on {@link DecoratedPayload}.
+   */
+  protected void handleDecoratedPayloadHeaders(final DecoratedPayload decoratedPayload,
+                                               final HttpServletResponse httpResponse)
+  {
+    // extra payload headers
+    final PayloadHeaders payloadHeaders = decoratedPayload.decorator(PayloadHeaders.class);
+    if (payloadHeaders != null) {
+      for (Map.Entry<String, String> header : payloadHeaders.getHeaders()) {
+        addHeaderIfNotPresent(
+            httpResponse,
+            header.getKey(),
+            header.getValue()
+        );
+      }
+    }
+    // times
+    final PayloadTimes payloadTimes = decoratedPayload.decorator(PayloadTimes.class);
+    if (payloadTimes != null) {
+      addHeaderIfNotPresent(
+          httpResponse,
+          HttpHeaders.LAST_MODIFIED,
+          payloadTimes.getLastModified()
+      );
+    }
+    // ETag using hashes
+    final PayloadHashes payloadHashes = decoratedPayload.decorator(PayloadHashes.class);
```

Comment:
This big method seems to make it really hard for format-specific senders to do custom thing - for example, maven might set the ETag differently.